androidApp/src/main/java/com/example/tasky/android/login/screen/LoginScreen.kt

```diff
@@ -1,23 +1,164 @@
 package com.example.tasky.android.login.screen
 
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.layout.Spacer
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.foundation.layout.height
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material3.Button
+import androidx.compose.material3.MaterialTheme.typography
 import androidx.compose.material3.Text
 import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.text.LinkAnnotation
+import androidx.compose.ui.text.SpanStyle
+import androidx.compose.ui.text.TextLinkStyles
+import androidx.compose.ui.text.buildAnnotatedString
+import androidx.compose.ui.text.withLink
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import androidx.compose.ui.unit.sp
 import androidx.navigation.NavGraphBuilder
 import androidx.navigation.compose.composable
-import com.example.tasky.Greeting
+import com.example.tasky.android.login.components.LoginPasswordTextField
+import com.example.tasky.android.login.components.LoginTextField
+import com.example.tasky.android.theme.Black
+import com.example.tasky.android.theme.Gray
+import com.example.tasky.android.theme.LightBlue
+import com.example.tasky.android.theme.MyApplicationTheme
 import kotlinx.serialization.Serializable
 
 fun NavGraphBuilder.loginScreen() {
     composable<Login> {
-        LoginScreen()
+        LoginScreen(
+            LoginScreenState(),
+            object : LoginScreenEvents {
+                override fun onClickToSignUp() {
+                }
+
+                override fun onClickLogin() {
+                }
+
+                override fun updateLoginScreenState(newState: LoginScreenState) {
+                }
+            },
+        )
     }
 }
 
 @Serializable
 object Login
 
+data class LoginScreenState(
+    val email: String = "",
+    val isEmailValid: Boolean = false,
+    val password: String = "",
+    val showPassword: Boolean = false,
+)
+
+interface LoginScreenEvents {
+    fun onClickToSignUp()
+
+    fun onClickLogin()
+
+    fun updateLoginScreenState(newState: LoginScreenState)
```

Comment:
Especially this function of the callback leaves the responsibility about how state should be updated to the UI. The idea behind UI and ViewModel is always to have a clear separation of tasks:
1. The UI should remain as "dumb" (as little logic) as possible and just observe state from the ViewModel. It then takes this state and decides how the UI needs to look like based on what it gets.
2. The ViewModel is the brain behind the UI which receives actions/events from the UI, like when the user clicks a button, and it then decides what should happen as a result of these actions to update the state.

The two main reasons behind this are (1) just a clearer structure & separation of concerns and (2) it makes it much easier to write unit tests for the actual logic behind your UI, since you wouldn't need to make these actual UI tests (which are very expensive and more error-prone), but you could just test your ViewModel.

Reply:
understand the points. but I have a question, especially when the states are all hoisted to ViewModels instead of the lowest common ancestor. Using a form as an example, it may contain more than 10 TextFields, and each TextField may have >= 4 states, there would be more than 40 state update events which is something that I want to avoid.

Reply:
I believe the states you are referring to is things like

1. `isError`
2. `isValid`
3. `value`
4. `onValueChange`

You could reduce this by making use of `TextFieldState` which handles state updates for you:

```
data class LoginState(
    val email: TextFieldState
    // …
)

// Usage in Compose
BasicTextField2(
    state = loginState.email
)
```

Regarding the other properties, these are things you want to be able to control from the ViewModel itself. You might need to perform validation on a different thread and when the result is ready, then set the textfield as valid or not. Think about if you need to check if an email address is registered on some remote server. 

Specifically for a Form screen, you could simplify the state by encapsulating some common properties:

```
data class FormTextFieldState(
    val state: TextFieldState = TextFieldState(),
    val isError: Boolean = false,
    val isValid: Boolean = false
```

And using them like this:

```
data class FormState(
    val email: FormTextFieldState = FormTextFieldState(),
    val username: FormTextFieldState = FormTextFieldState()
    // …
)
```